Add new movement based achievements and statistics

ralreegorganon opened the pull request:
#### Summary

SUMMARY: Content "Add new movement based achievements and statistics"

#### Purpose of change

I wanted to take the new achievements and statistics systems for a spin.

#### Describe the solution

Followed the docs and examples to create new elements that create event streams for player movement, including walking, running, crouching, swimming, and swimming underwater, and then used them in scores and achievements.

Updated the value_constraint for cata_variant to support a boolean.

Wrote some new event field transformations, including converting an overmap terrain id into a collection of flags, a characters mount id into a boolean for whether or not they were mounted, and the terrain into a boolean for whether or not movement in it would be swimming.

![image](https://user-images.githubusercontent.com/11464/80018801-0fddd580-8483-11ea-8b2f-da83bcb4e0c2.png)

![image](https://user-images.githubusercontent.com/11464/80018812-1409f300-8483-11ea-9e4f-581afccec4d7.png)

#### Testing

Unit tests. Played the game to confirm stats and achievements.

#### Additional context

I split the various JSON types involved in this into three files: achievements, scores, and statistics. They don't have to be, but I was finding it a bit confusing in the single file scrolling back and forth everything, and felt like this groups them into categories that make sense to me--achievements and scores are both independent user-facing things, while the statistics is the grab-bag of `event_statistic` and `event_transformation` objects that are used by everything. If someone is really opposed to this split, it's not a big deal to move them back, but I think they're going to get even harder to navigate as we add more.

I'm not particularly attached to these achievement definitions or names--I just wanted to exercise the systems and add some content. I could've written more, but this seemed a decent start for trying things out and I didn't want this pull request to get any bigger than it already is.

@jbytheway I'm leaving this in draft because I need some help figuring out how we want to handle modifications to existing events in existing games. When this runs against an existing previous save, we end up triggering an error on the event transformation that is reading `avatar_moves` and and trying to apply the new transformations to fields that didn't exist before. The game continues just fine, but it seems pretty egregious to trigger this user facing error every time you go to the stats screen. Some options that come to mind:
1. Just give the score for moves walked a new name (painful as the inconsistency may be) and delete `score_walked` entirely (the rest of the pipeline between `avatar_moves` and `score_walked` is already gone with the deletion of `moves_not_mounted`).
2. Deprecate/version `avatar_moves` and make an entirely new version with the new properties and adjust all of the event calls in the C++ and everything downstream in the pipeline.
3. Remove the debugmsg in favor of something that doesn't interrupt the user (the problem being that deliberately missing data is indistinguishable from a typo in the key in the transformation).
4. Something else? I'm kinda out of ideas here and don't really want to do something elaborate for something that only applies to old saves for a fun but non-critical path feature, but it also seems like it's worth figuring out now before we have hundreds of events and statistics...

jbytheway commented on src/event_statistics.cpp:
```diff
@@ -178,10 +178,15 @@ struct value_constraint {
     void deserialize( JsonIn &jsin ) {
         JsonObject jo = jsin.get_object();
         int equals_int;
+        bool equals_bool;
         if( jo.read( "equals", equals_int, false ) ) {
             equals_ = cata_variant::make<cata_variant_type::int_>( equals_int );
         }
 
+        if( jo.read( "equals", equals_bool, false ) ) {
```
```suggestion
        if( jo.read( "equals", equals_int, false ) ) {
            equals_ = cata_variant::make<cata_variant_type::int_>( equals_int );
        }

        bool equals_bool;
        if( jo.read( "equals", equals_bool, false ) ) {
```

ralreegorganon replied:
Done.

jbytheway commented on src/event_field_transformations.cpp:
```diff
@@ -17,6 +18,36 @@ static std::vector<cata_variant> species_of_monster( const cata_variant &v )
+}
+
 const std::unordered_map<std::string, EventFieldTransformation> event_field_transformations = {
     { "species_of_monster", species_of_monster },
+    { "terrain_flag", terrain_flag },
+    { "mount_to_mounted", mount_to_mounted },
+    { "terrain_to_swimming", terrain_to_swimming },
```
Suggest we keep these in alphabetical order, for want of a reason to do otherwise.

ralreegorganon replied:
Done.

jbytheway commented on src/avatar_action.cpp:
```diff
@@ -562,6 +562,14 @@ void avatar_action::swim( map &m, avatar &you, const tripoint &p )
+    if( you.is_mounted() ) {
+        mount_type = you.mounted_creature->type->id;
+    }
+    g->events().send<event_type::avatar_moves>( mount_type, m.ter( p ).id(),
+            you.get_movement_mode_str(), you.is_underwater() );
```
Thanks; I was wanting to add this one myself, and I hadn't found the place to do it yet :).

jbytheway commented:
Regarding handling of old saves:

* I've already set things up so that scores which didn't exist when you start a game don't appear in the stats screen, so we don't have to worry about newly created scores being calculated incorrectly.
* I think in this case you've changed the way an existing score is computed.  That's a trickier case.  I think we have to make a decision on a case-by-case basis when that happens whether we should change its id (thus removing it from ongoing games) because it would be calculated incorrectly.
* I plan to add proper type checking at "Verifying" stage to all the json-defined entities here.  Once we have that it will be less important if we don't report errors related to missing fields during the game.  I'll try to make that the next thing I implement in support of achievements.
* Another option might be to mark events that were loaded from the save in some way, expect those ones to be potentially out of date, and suppress errors related to them.

ZhilkinSerg commented on src/event.h:
```diff
@@ -168,8 +168,11 @@ struct event_spec<event_type::angers_amigara_horrors> : event_spec_empty {};
+    static constexpr std::array<std::pair<const char *, cata_variant_type>, 4> fields = {{
             { "mount", cata_variant_type::mtype_id },
+            { "terrain", cata_variant_type::ter_id },
+            { "movement_mode", cata_variant_type::string },
```
Maybe #38299 will be finished someday

ralreegorganon replied:
Sure thing.

ralreegorganon replied:
Done. Needed to do a bit more boilerplate on that enum (e.g. it didn't have an enum_to_string implementation) but it's good now.

ralreegorganon commented on data/json/scores/achievements.json:
```diff
@@ -0,0 +1,32 @@
+[
```
No big deal, I'll resolve it after yours is merged--I won't even take this out of draft until then.

ralreegorganon commented on data/json/scores/achievements.json:
```diff
@@ -0,0 +1,32 @@
+  {
+    "id": "achievement_swim_short_distance",
+    "type": "achievement",
+    "description": "A Swimming Merit Badge",
```
Yeah, I thought the same but only had your single example so far. I don't have a real strong opinion either--I just picked the title case because that's the convention that I see everywhere. Your system, you decide and I'll update 😄 .

jbytheway replied:
I've looked at Steam and it appears that you're right.  Almost everyone seems to be capitalizing every word.  But Factorio doesn't, and that's clearly the most important game to imitate ;).

Ultimately I think it comes down to what style of name we're using.  If they were mostly one to three word names, then maybe All Initial Caps would work, but we seem to be leaning towards longer names, where I think it starts to look silly.  So I'm voting for not doing that.   Just use regular sentence case.

ralreegorganon commented on data/json/scores/achievements.json:
```diff
@@ -0,0 +1,32 @@
+    "id": "achievement_marathon",
+    "type": "achievement",
+    "description": "Pheidippides Was A Hack",
+    "requirements": [ { "event_statistic": "num_moves_ran", "is": ">=", "target": 42196 } ]
```
Yup, distance of a marathon +1 meter. I also agree on the name / description differentiation, because I actually toyed with putting that "run a marathon" into the description here directly with the name, but that looked terrible, and I really didn't want to get into reworking the UI part of this (and I'm glad I didn't because I didn't see your achievement PR that updates it until this morning). It'd be good to have the name, narrative description, and then the statistical counters.

ralreegorganon commented on data/json/scores/scores.json:
```diff
@@ -0,0 +1,68 @@
+  {
+    "id": "score_kills",
+    "type": "score",
+    "description": "Monsters killed: %s",
```
Yeah, it felt a little redundant for a bunch of these. I'll wait for #39686 to land and then update.

I made the scores I added (and the existing ones) more concise than they were before, e.g. `Monsters killed: %s` rather than `Number of monsters killed: %s` because the original felt too long, and the only different between the score and the stat was just the leading capital, which I can easily update the stat to have and drop the score description.

ralreegorganon commented:
> Regarding handling of old saves:

> I think in this case you've changed the way an existing score is computed. That's a trickier case. I think we have to make a decision on a case-by-case basis when that happens whether we should change its id (thus removing it from ongoing games) because it would be calculated incorrectly.

I think this one is probably the salient point--because I retained the id but changed the pipeline for the score, including having new fields get transformed that didn't even exist previously, it breaks. I think I'm probably gonna bite the bullet and just assign a new id for this score, and accept the inconsistency in movement mode statistic ids (or maybe make them all consistent with however I name the new one). 

The thing I'm still not sure about is how we want to handle this replacement even if I do that. I don't really want to just keep the old score around because then it would continue to be used in new games, and I was thinking the new stats should replace it rather than co-exist with the old one. Do we need some sort of attribute that says "you can load this if it exists in your game but do NOT load this in a new game?" and then at some point we can nuke the deprecated pipeline?

I supposed I could just keep the entire old stat and pipeline for it and just rename the user facing part of it to more accurately reflect what it is--it's not tiles walked, but rather tiles traveled unassisted by a mount or vehicle (I could call it "on foot" though it might be a little funny since it'd include swimming). That might be the easiest since fiddling with this is a lot less fun than writing new stats and achievements. 😄

jbytheway commented:
> The thing I'm still not sure about is how we want to handle this replacement even if I do that. I don't really want to just keep the old score around because then it would continue to be used in new games, and I was thinking the new stats should replace it rather than co-exist with the old one. Do we need some sort of attribute that says "you can load this if it exists in your game but do NOT load this in a new game?" and then at some point we can nuke the deprecated pipeline?

Maybe we'll need such a feature at some point, but for now I don't think anyone is too invested in their scores, so I suggest just delete the old one and use a new id for the new one.  People with ongoing games will get neither the new nor the old one, but I don't think they'll care.

jbytheway commented:
>     * I plan to add proper type checking at "Verifying" stage to all the json-defined entities here.  Once we have that it will be less important if we don't report errors related to missing fields during the game.  I'll try to make that the next thing I implement in support of achievements.

I've implemeted this now: additional checking at verification stage, so that we can safely ignore errors later, because they ought to only arise from stale data from the save.  I'll PR it once #39686 is merged, since it builds on that.  You'll probably want to wait until after both are merged to rebase this branch.

ralreegorganon commented:
> I've implemented this now: additional checking at verification stage, so that we can safely ignore errors later, because they ought to only arise from stale data from the save. I'll PR it once #39686 is merged, since it builds on that. You'll probably want to wait until after both are merged to rebase this branch.

Excellent--no big deal on rebasing. I've actually got another branch I've started off of this one because I wanted to add `maximum` and `minimum` stat types, so it's just one more rebase for the pile.

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -567,12 +567,7 @@ void Character::load( const JsonObject &data )
-        if( tmove_mode == character_movemode_str[i] ) {
-            move_mode = static_cast<character_movemode>( i );
-            break;
-        }
-    }
+    move_mode = io::string_to_enum<character_movemode>( tmove_mode );
```
The JSON reading code has the enum string conversion built-in; this shouldn't need the intermediate string variable; you can just read the enum directly.

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -713,7 +708,7 @@ void Character::store( JsonOut &json ) const
     json.member( "my_bionics", *my_bionics );
 
-    json.member( "move_mode", character_movemode_str[ move_mode ] );
+    json.member( "move_mode", io::enum_to_string( move_mode ) );
```
```suggestion
    json.member_as_string( "move_mode", move_mode );
```

jbytheway commented on src/avatar_action.cpp:
```diff
@@ -562,6 +562,14 @@ void avatar_action::swim( map &m, avatar &you, const tripoint &p )
     you.setpos( p );
     g->update_map( you );
+
+    mtype_id mount_type;
+    if( you.is_mounted() ) {
+        mount_type = you.mounted_creature->type->id;
+    }
+    g->events().send<event_type::avatar_moves>( mount_type, m.ter( p ).id(),
+            you.get_movement_mode(), you.is_underwater() );
+
```
This chunk of code is duplicated now; should probably be pulled out into a function.

jbytheway commented on src/event_field_transformations.cpp:
```diff
@@ -17,6 +25,29 @@ static std::vector<cata_variant> species_of_monster( const cata_variant &v )
     return result;
 }
 
+static std::vector<cata_variant> terrain_flag( const cata_variant &v )
```
I would name this `flags_of_terrain`.  Then when it's used like
```
"new_fields": { "terrain_flag": { "flags_of_terrain": "terrain" } }
```
the concepts are near each other visually: "flag" near "flag" and "terrain" near "terrain".

jbytheway commented on src/event_field_transformations.cpp:
```diff
@@ -2,10 +2,18 @@
 #include "string_id.h"
 #include "type_id.h"
 
+static std::vector<cata_variant> mount_to_mounted( const cata_variant &v )
```
This could be a generic `is_nonempty` that works on an arbitrary variant by examining the string directly.  That's probably a good idea, although it conflicts a little with (unpublished) changes I made to improve type safety.

jbytheway commented on src/event_field_transformations.cpp:
```diff
@@ -17,6 +25,29 @@ static std::vector<cata_variant> species_of_monster( const cata_variant &v )
+    return result;
+}
+
+static std::vector<cata_variant> terrain_to_swimming( const cata_variant &v )
```
I'd probably call this `is_swimming_terrain`.  Naming something with a `is_` name is a nice way to signal that it returns bool.

ralreegorganon commented:
> Some more review comments. Some of these are more subjective, so feel free to disagree.

These all make sense. I've made all the changes except for the generic `is_nonempty`. 

I'd rather wait on that one both because I'd like to have your unpublished type safety changes in place first and because I'm not entirely certain about the best path forward. That default constructed mtype_id has an empty string, but an empty string is totally valid for the `string_id<T>`,  and we have the ::NULL_ID() for the absence of a value but that's not being used. Regardless, I do think some generic implementation of "this cata_variant does not have a value" would be useful.

aoae23 commented on data/json/scores/achievements.json:
```diff
@@ -0,0 +1,28 @@
+  {
+    "id": "achievement_traverse_sharp_terrain",
+    "type": "achievement",
+    "description": "Every rose has its thorn",
```
thorns*

jbytheway commented:
Next batch of functionality now PRed in #39777.  Does that resolve the issues with getting error messages for old saves?

ralreegorganon commented:
> Next batch of functionality now PRed in #39777. Does that resolve the issues with getting error messages for old saves?

Nope. I had to recreate the issue since I'd already resolved it here by just making a new score and deleting the old one. To do so, I rebased this PR on that, implemented the new definitions for the event_field_transformations that list the expected inputs and outputs, and then renamed my "score_moves_walked" back to "score_walked" and then loaded a save where the "score_walked" existed and it still gives an error `"Expected input field 'terrain' not present in event"`.

The old pipeline was:
score_walked -> num_moves_not_mounted -> moves_not_mounted -> avatar_moves

The new one to recreate the error was:
score_walked -> num_moves_walked -> moves_walked ->moves_expand_pre_move_mode -> avatar_moves

jbytheway commented:
Good job we checked.  Thanks.  Looks like I missed one `debugmsg`.  I've pushed a new version to #39777 now which hopefully fixes it.

ralreegorganon commented:
> Good job we checked. Thanks. Looks like I missed one debugmsg. I've pushed a new version to #39777 now which hopefully fixes it.

Rebased and it's working as expected now. 👍

jbytheway commented on data/json/achievements.json:
```diff
@@ -11,5 +11,25 @@
+    "requirements": [ { "event_statistic": "num_moves_ran", "is": ">=", "target": 42196 } ]
+  },
+  {
+    "id": "achievement_proclaimers",
```
I'd give this a name closer to the definition rather than the inspiration, like `achievement_walk_1000_miles`.  Just in case it ever changes description.

ralreegorganon replied:
Fine by me.
